torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1190,25 +1246,45 @@ class ShapePropagator {
     //   - has bool keepdim and int[] dim arguments
     static const register_formula_for multidim_reduce_ops_with_integer_upcast{
         {
-            "aten::sum(Tensor self, int[] dim, bool keepdim) -> Tensor",
+            "aten::sum(Tensor self, int[] dim, bool keepdim, *, int? dtype) -> Tensor",
+            "aten::mean(Tensor self, int[] dim, bool keepdim, *, int? dtype) -> Tensor",
         },
         [](Node* node) -> type_vec_t {
-          if (auto dim = node->get<std::vector<int64_t>>(attr::dim)) {
+          if (auto dims = node->namedInput(attr::dim)->node()->inputs().size()) {
+            auto opt_keepdim = node->get<bool>(attr::keepdim);
+            at::optional<IValue> opt_dtype = node->get(attr::dtype);
+            auto num_reduced = *opt_keepdim ? 0 : dims;
             // TODO: can dim contain duplicates?
-            return multidim_reduce_with_postprocess(
-                node, /*num_reduced_dim=*/dim->size(), /*upcast_integer=*/true);
+            return reduce_op_handler(
+                node, /*num_reduced_dim=*/num_reduced, /*upcast_integer=*/true, opt_dtype);
```

Comment:
prior code had upcast=false for mean, but it's not necessary to differentiate since mean isn't implemented for integers anyway.